File: Server/src/main/core/network/packet/outgoing/GrandExchangePacket.java
package core.network.packet.outgoing;

import core.network.packet.IoBuffer;
import core.network.packet.OutgoingPacket;
import core.network.packet.PacketHeader;
import core.network.packet.context.GrandExchangeContext;

/**
 * The outgoing packet used for updating a player's grand exchange data.
 * @author Emperor
 * @author Vexia
 * @author Angle
 */
public class GrandExchangePacket implements OutgoingPacket<GrandExchangeContext> {

	private final int REMOVED = 6;
	private final int ABORTED = 2;

	@Override
	public void send(GrandExchangeContext context) {
		final IoBuffer buffer = new IoBuffer(116, PacketHeader.NORMAL);
		buffer.put(context.idx);
		if (context.state == REMOVED) {
			buffer.put(0).putShort(0).putInt(0).putInt(0).putInt(0).putInt(0);
		} else {
			byte state = (byte) (context.state + 1);
			if (context.isSell) {
				state += 8;
			}
			if (context.state == ABORTED) {
				state = context.isSell ? (byte) -3 : (byte) 5;
			}
			buffer.put(state).putShort(context.itemID).putInt(context.value).putInt(context.amt).
					putInt(context.completedAmt).putInt(context.totalCoinsExchanged);
		}
		try {
			buffer.cypherOpcode(context.getPlayer().getSession().getIsaacPair().getOutput());context.getPlayer().getSession().write(buffer);
		} catch (Exception e) {
		}
	}

}

File: Server/src/main/core/network/packet/outgoing/GrandExchangePacket.kt
package core.network.packet.outgoing

import core.network.packet.IoBuffer
import core.network.packet.OutgoingPacket
import core.network.packet.PacketHeader
import core.network.packet.context.GrandExchangeContext

class GrandExchangePacket : OutgoingPacket<GrandExchangeContext> {
    private val removed = 6
    private val aborted = 2

    override fun send(context: GrandExchangeContext) {
        val buffer = IoBuffer(116, PacketHeader.NORMAL)
        buffer.put(context.idx.toInt())
        if (context.state.toInt() == removed) {
            buffer.put(0).putShort(0).putInt(0).putInt(0).putInt(0).putInt(0)
        } else {
            var state = (context.state + 1).toByte()
            if (context.isSell) {
                state = (state + 8).toByte()
            }
            if (context.state.toInt() == aborted) {
                state = if (context.isSell) (-3).toByte() else 5.toByte()
            }
            buffer.put(state.toInt()).putShort(context.itemID.toInt()).putInt(context.value).putInt(context.amt)
                .putInt(context.completedAmt).putInt(context.totalCoinsExchanged)
        }
        try {
            buffer.cypherOpcode(context.player.session.isaacPair.output)
            context.player.session.write(buffer)
        } catch (e: Exception) {
            // Do nothing
        }
    }
}

